File: 2016/Day6/main.lua
local function decode_message(messages, find_least_frequent)
    if #messages == 0 then return '' end

    local num_columns = #messages[1]
    local frequency_maps = {}

    for i = 1, num_columns do
        frequency_maps[i] = {}
    end

    for _, message in ipairs(messages) do
        for i = 1, num_columns do
            local char = message:sub(i, i)
            frequency_maps[i][char] = (frequency_maps[i][char] or 0) + 1
        end
    end

    local decoded_message = ''
    for i = 1, num_columns do
        local frequency_map = frequency_maps[i]
        local selected_char = ' '
        local selected_frequency = find_least_frequent and math.huge or 0

        for char, frequency in pairs(frequency_map) do
            if find_least_frequent and frequency < selected_frequency or
               not find_least_frequent and frequency > selected_frequency then
                selected_char = char
                selected_frequency = frequency
            end
        end

        decoded_message = decoded_message .. selected_char
    end

    return decoded_message
end

local messages = {}
for line in io.lines('input.txt') do
    table.insert(messages, line)
end

print('Part 1: ' .. decode_message(messages, false))
print('Part 2: ' .. decode_message(messages, true))

File: 2016/Day6/main.py
from collections import Counter
import sys


def decode_message(messages: list[str], find_least_frequent: bool = False) -> str:
    if not messages: return ''

    num_columns = len(messages[0])
    frequency_maps: list[Counter[str]] = [Counter() for _ in range(num_columns)]

    for message in messages:
        for i in range(num_columns):
            frequency_maps[i][message[i]] += 1

    decoded_message = []
    for frequency_map in frequency_maps:
        if not frequency_map:
            decoded_message.append('')
            continue

        if find_least_frequent:
            selected_char = min(frequency_map.items(), key=lambda item: item[1])[0]
        else:
            selected_char = max(frequency_map.items(), key=lambda item: item[1])[0]

        decoded_message.append(selected_char)

    return ''.join(decoded_message)

if __name__ == '__main__':
    messages = [line.strip() for line in sys.stdin if line.strip()]

    print(f'Part 1: {decode_message(messages)}')
    print(f'Part 2: {decode_message(messages, True)}')
